test: User Controller, Service, Domain 테스트 코드 작성

c-yeonwoo opened the pull request:
## Resolve #44

- 유저 인증과 유저 정보를 조회하는 코드에 대한 테스트가 필요하다.

## Changes

- UserControllerTest, UserServiceTest, UserTest 클래스를 만들고 각각에 단위테스트를 추가했다.

## Notes

- N/A

## References

- [JUnit5 소개 - 자바캔 블로그](https://javacan.tistory.com/entry/JUnit-5-Intro)

c-yeonwoo committed:
test: UserController의 findUser 단위 테스트 추가

c-yeonwoo committed:
test: UserService saveOrUpdate 단위 테스트 추가

c-yeonwoo committed:
test: UserService findById 단위 테스트 추가

c-yeonwoo committed:
test: User updateUserIno 단위 테스트 추가

sonypark commented on src/test/java/underdogs/devbie/user/domain/UserTest.java:
```diff
@@ -0,0 +1,53 @@
+
+import underdogs.devbie.auth.dto.UserInfoResponse;
+
+@SpringBootTest
```
UserTest는 도메인 모델 테스트라 Spring Container가 필요하지 않으니 @SpringBootTest 어노테이션은 안 써도 되지 않을까요?😊

@SpringBootTest 어노테이션은 스프링 컨테이너에 어플리케이션의 모든 Bean을 올려서 테스트 속도가 많이 느리다 합니다. 그래서 단위 테스트에서는 @SpringBootTest 사용을 지양하는게 좋을 것 같습니다.

[Integration Tests with @SpringBootTest](https://reflectoring.io/spring-boot-test/)

Spring Boot provides the @SpringBootTest annotation which we can use to create an application context containing all the objects we need for all of the above test types. Note, however, that overusing @SpringBootTest might lead to very long-running test suites.

So, for simple tests that cover multiple units we should rather create plain tests, very similar to unit tests, 


[Optimizing Spring Integration Tests | Baeldung](https://www.baeldung.com/spring-tests)

Integration tests help us build confidence but they come at a price:

- That is a slower execution speed, which means slower builds
- Also, integration tests imply a broader testing scope which is not ideal in most cases

c-yeonwoo replied:
감사합니다. 
저도 소니의견에 동의해요!

serivce와 domain의 단위테스트는 @SpringBootTest 어노테이션을 제거하고
controller에서는 @WebMvcTest로 수정 반영해놓겠습니다!

sonypark commented on src/test/java/underdogs/devbie/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,74 @@
+import underdogs.devbie.user.domain.User;
+import underdogs.devbie.user.domain.UserRepository;
+
+@SpringBootTest
```
UserService 테스트도 UserRepository를 @Mock으로 대체하니 @SpringBootTest 없이 테스트 가능할 것 같습니다 :)

sonypark commented on src/test/java/underdogs/devbie/user/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,56 @@
+import underdogs.devbie.auth.controller.resolver.LoginUserArgumentResolver;
+import underdogs.devbie.user.domain.User;
+
+@SpringBootTest
```
Controller 테스트도 @SpringBootTest가 아닌 **@WebMvcTest**로도 가능할 것 같은데, 이 부분은 저도 한 번 확인해봐야겠네요 🤔

sonypark reviewed (changes requested):
보스독 테스트 깔끔하게 잘 짜셨네요.
덕분에 저도 공부 많이 됐습니다.
간단한 리뷰 남겼는데 확인해보시고 수정 가능한 부분은 반영해주시면 감사하겠습니다!

팀원들과 논의해보면 좋을 내용
- @Mock vs @MockBean 차이
- @SpringBootTest 를 언제 사용할것인지
  - 꼭 사용해야할 경우
  - 사용하지 않아도 괜찮은경우

KimSeongGyu1 commented on src/test/java/underdogs/devbie/user/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,56 @@
+    @DisplayName("유저 정보 조회")
+    @Test
+    void findUser() throws Exception {
+        // given
+        User user = User.builder()
+            .id(1L)
+            .oauthId(TEST_OAUTH_ID)
+            .email(TEST_USER_EMAIL)
+            .build();
+        given(bearerAuthInterceptor.preHandle(any(), any(), any())).willReturn(true);
+        given(loginUserArgumentResolver.supportsParameter(any())).willReturn(true);
+        given(loginUserArgumentResolver.resolveArgument(any(), any(), any(), any())).willReturn(user);
+
+        // when
+        mockMvc.perform(get("/api/user")
+            .header("Authorization", "bearer JWT_ACCESS_TOKEN")
+            .contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isOk())
+            .andExpect(content().string(containsString("\"email\":\"underdogs@devbie.link\"")));
```
저는 //given  //when을 명시하지 않고 했는데 저와 다르게 명시해 주셨네요
//given //when 을 명시적으로 표시해 줄 것인가에 대해서 팀원 모두가 같이 이야기해 봐야 할 것 같네요

KimSeongGyu1 replied:
그리고 저는 mockMvc.perform()이 자주 사용될 것 같아 따로 분리해서 메서드화 했는데
이 부분은 작업 영역이 겹칠 것 같으니 내일 어떻게 충돌 관리할 지 같이 이야기 해야 할 것 같아요

hongbin-dev replied:
@KimSeongGyu1 
mockMvc 메서드를 편하게 사용하도록 메서드추출하자는 말씀인가요??

slowCoyle replied:
given, when, then이 큰 비용을 발생시키는 것이 아닌 만큼 지향하면 좋겠다고 생각합니다! :)

KimSeongGyu1 commented on src/test/java/underdogs/devbie/user/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,56 @@
+            .oauthId(TEST_OAUTH_ID)
+            .email(TEST_USER_EMAIL)
+            .build();
+        given(bearerAuthInterceptor.preHandle(any(), any(), any())).willReturn(true);
+        given(loginUserArgumentResolver.supportsParameter(any())).willReturn(true);
+        given(loginUserArgumentResolver.resolveArgument(any(), any(), any(), any())).willReturn(user);
```
예전에 미션 진행할 때는 any()를 자주 사용했는데 이 부분이 문제를 야기할 수 있을까요?

KimSeongGyu1 commented on src/test/java/underdogs/devbie/user/domain/UserTest.java:
```diff
@@ -0,0 +1,53 @@
+            .oauthId(TEST_OAUTH_ID)
+            .email(TEST_USER_EMAIL)
+            .build();
+        UserInfoResponse userInfoResponse = new UserInfoResponse(TEST_OAUTH_ID, "Changed Email");
```
이 부분에서 한번만 쓰일지도 모르지만
changed Email도 상수화를 하면 어떨까요

KimSeongGyu1 commented on src/test/java/underdogs/devbie/user/domain/UserTest.java:
```diff
@@ -0,0 +1,53 @@
+    @Test
+    void validateUserInfo() {
+        // given
+        User user = User.builder()
+            .id(1L)
+            .oauthId(TEST_OAUTH_ID)
+            .email(TEST_USER_EMAIL)
+            .build();
```
모든 테스트에서 같은 User를 매번 생성하고 있네요
서비스 테스트처럼 개선하면 문제가 있을까요

KimSeongGyu1 commented on src/test/java/underdogs/devbie/user/service/UserServiceTest.java:
```diff
@@ -0,0 +1,74 @@
+
+        // then
+        verify(userRepository).findByOauthId(TEST_OAUTH_ID);
+        assertThat(user.getId()).isEqualTo(1L);
+        assertThat(user.getOauthId()).isEqualTo(TEST_OAUTH_ID);
+        assertThat(user.getEmail()).isEqualTo(TEST_USER_EMAIL);
```
여러 개의 assert 문이 한번의 Test안에 공존하지 않는 방향으로 하는 것이 좋다고 들은 적이 있는데
어떻게 생각하시나요
다른 팀원분들도 이야기 해줬으면 좋겠어요

hongbin-dev replied:
User 상태를 검증하려는 부분인 것 같은데 유안의 피드백을 수용하면

```
User expect = new User(...)
assertThat(user).isEqualTo(expect)
```
이런식으로 대체가 되겠네요. 
저는 1개의 객체의 상태를 검증하는데는 여러개의 assertThat()을 사용해도 괜찮다고 생각합니다.

slowCoyle replied:
하나의 Test안에 여러 개의 이슈?(개념?)을 테스트하는 것은 안좋다는 얘기는 들어봤습니다.
즉, 하나의 Test는 하나의 이슈만을 테스트해야한다는 개념인데요.

여러 개의 assert문이 문제가 될 수 있다는 소리는 처음 접해보네요!
혹시 어떤 이유로 여러 assert문이 공존하지 않는 것이 방향이 좋을까요?

sonypark replied:
한 테스트 메서드에서 assertThat()를 여려번 사용할 경우 assertAll()을 이용하는게 좋다고 생각합니다.
그 이유는 assertThat()을 연달아 사용할 경우 이전 assertThat()이 fail일 경우 그 다음 선언된 assertThat()은 호출 되지 않기 때문입니다.
assertAll()을 사용하면 각각의 assertThat()의 성공, 실패 결과를 모두 알 수 있어 테스트 결과를 확인하거나 디버깅할 때 편리할 것 같습니다.

관련 내용은 [stackoverflow](https://stackoverflow.com/questions/40796756/assertall-vs-multiple-assertions-in-junit5)에도 잘 나와있으니 참고하시면 좋을 것 같아요!

단, 필드의 값을 하나하나 확인하는 경우해야 하는 경우가 아니라면 하나의 테스트에는 하나의 assert 문만 사용하는걸 지향할 필요가 있다고 생각합니다.
But do not abuse it! A single test method should always test only one assumption about the production code. That's the main reason why you usually have only one assert per test method

KimSeongGyu1 reviewed (commented):
코드작업이 끝나고 보니 
보스독이 제일 고생 많이 한 것 같네요
고생하셨어요

코멘트들은 팀원 모두 같이 생각해봤으면 좋겠다고 생각한 점 위주로 남겼어요
같이 생각해봐요

hongbin-dev commented on src/test/java/underdogs/devbie/user/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,56 @@
+        given(loginUserArgumentResolver.supportsParameter(any())).willReturn(true);
+        given(loginUserArgumentResolver.resolveArgument(any(), any(), any(), any())).willReturn(user);
+
+        // when
+        mockMvc.perform(get("/api/user")
+            .header("Authorization", "bearer JWT_ACCESS_TOKEN")
+            .contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isOk())
+            .andExpect(content().string(containsString("\"email\":\"underdogs@devbie.link\"")));
+    }
```
위 작업은 when과 then을 동시에 수행하는 작업이 아닐까요..?

hongbin-dev reviewed (commented):
보스독~ 앨런입니다.
이번 주 스프린트하면서 정말 고생많았어요~!
간단한 코드리뷰 남겼으니 확인부탁드려요
감사합니다..!👍👍

c-yeonwoo committed:
test: test class annotation 조정

service, domain -> annotation 제거
controller -> @SpringBootTest 대신 @WebMvcTest사용

c-yeonwoo committed:
test: controller class annotation 조정

controller -> @WebMvcTest에 controller class 명시

c-yeonwoo committed:
test: BDD 주석 제거 및 assertAll 적용

c-yeonwoo committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-devbie into test/44-user_test

c-yeonwoo committed:
test: MvcTest 상속 적용, 상수 추출

c-yeonwoo committed:
refactor: AuthController request url 변경

/api/oauth -> /api/auth
